test/functional/test_framework/v2_p2p.py

```diff
@@ -4,17 +4,280 @@
+        return -1
+
+    def complete_handshake(self, response):
+        """ Instantiates the encrypted transport and
+        sends garbage terminator + optional decoy packets + transport version packet.
+        Done by both initiator and responder.
+
+        Returns:
+        bytes - bytes to be sent to the peer when completing the v2 handshake
+        """
+        received_prefix = b'' if self.initiating else self.received_prefix
+        ellswift_theirs = received_prefix + response.read(64 - len(received_prefix))
+        ecdh_secret = self.v2_ecdh(self.privkey_ours, ellswift_theirs, self.ellswift_ours, self.initiating)
+        self.initialize_v2_transport(ecdh_secret)
+        # Send garbage terminator
+        msg_to_send = self.peer['send_garbage_terminator']
+        # Optionally send decoy packets after garbage terminator.
+        aad = self.sent_garbage
+        for decoy_content_len in [random.randint(1, 100) for _ in range(random.randint(0, 10))]:
+            msg_to_send += self.v2_enc_packet(decoy_content_len * b'\x00', aad=aad, ignore=True)
+            aad = b''
+        # Send version packet.
+        msg_to_send += self.v2_enc_packet(TRANSPORT_VERSION, aad=aad)
+        return msg_to_send
+
+    def authenticate_handshake(self, response):
+        """ Ensures that the received optional decoy packets and transport version packet are authenticated.
+        Marks the v2 handshake as complete. Done by both initiator and responder.
+
+        Returns:
+        1. int - length of bytes that were processed so that recvbuf can be updated
+        2. bool - True if the authentication was successful/more bytes need to be received and False otherwise
+        """
+        received_garbage = response[:16]
+        response = response[16:]
+        processed_length = len(received_garbage)
+        for i in range(MAX_GARBAGE_LEN + 1):
+            if received_garbage[-16:] == self.peer['recv_garbage_terminator']:
+                # Receive, decode, and ignore version packet.
+                # This includes skipping decoys and authenticating the received garbage.
+                aad = received_garbage[:-16]
+                while not self.tried_v2_handshake:
+                    length, contents = self.v2_receive_packet(response, aad=aad)
+                    aad = b""
+                    if length == -1:
+                        return processed_length, False
+                    elif length == 0:
+                        return 0, True
+                    processed_length += length
+                    # decoy packets have contents = None. v2 handshake is complete only when version packet
+                    # (can be empty with contents = b"") with contents != None is received.
+                    if contents is not None:
+                        self.tried_v2_handshake = True
+                        return processed_length, True
+                    response = response[length:]
```

Comment:
just want to add that if this turned out annoying to fix (I didn't see an immediate trivial fix but haven't looked hard), I'd be fine just throwing an assert / error if decoy packages are received in this PR (since bitcoind shouldn't send those anyway).

Reply:
great catch! i've changed the `authenticate_handshake` function to first deal with garbage terminator detection before processing decoy packets and version packet. hopefully this is not a concern now?